packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -144,32 +145,46 @@ export class MonacoWorkspace extends BaseMonacoWorkspace implements lang.Workspa
         };
     }
 
-    applyEdit(changes: lang.WorkspaceEdit): Promise<boolean> {
+    async applyEdit(changes: lang.WorkspaceEdit): Promise<boolean> {
         const workspaceEdit = this.p2m.asWorkspaceEdit(changes);
-        const promises = [];
-        for (const edit of workspaceEdit.edits) {
-            promises.push(this.textModelService.createModelReference(edit.resource).then(reference => {
-                const model = reference.object.textEditorModel;
-                // start a fresh operation
-                model.pushStackElement();
-                const range = edit.range;
-                const selections = [new monaco.Selection(range.startLineNumber, range.startColumn, range.endLineNumber, range.endColumn)];
-                model.pushEditOperations(selections, [{
+        const uri2Edits = this.groupEdits(workspaceEdit);
+        for (const uri of uri2Edits.keys()) {
+            const editorWidget = await this.editorManager.open(new URI(uri));
+            const editor = get(editorWidget);
+            if (editor) {
+                const model = editor.document.textEditorModel;
+                const currentSelection = new monaco.Selection(
+                    editor.selection.start.line,
+                    editor.selection.start.character,
+                    editor.selection.end.line,
+                    editor.selection.end.character);
+                const edits = uri2Edits.get(uri)!;
+                const editOperations: monaco.editor.IIdentifiedSingleEditOperation[] = edits.map(edit => ({
                     identifier: undefined!,
                     forceMoveMarkers: false,
-                    range: new monaco.Range(range.startLineNumber, range.startColumn, range.endLineNumber, range.endColumn),
+                    range: new monaco.Range(edit.range.startLineNumber, edit.range.startColumn, edit.range.endLineNumber, edit.range.endColumn),
                     text: edit.newText
-                }], (edits) => selections);
-                const editor = this.editorManager.editors.find(editor => editor.editor.uri.toString() === model.uri.toString());
-                if (editor) {
-                    editor.editor.focus();
-                }
+                }));
+                // start a fresh operation
+                model.pushStackElement();
+                model.pushEditOperations([currentSelection], editOperations, (undoEdits: monaco.editor.IIdentifiedSingleEditOperation[]) => [currentSelection]);
                 // push again to make this change an undoable operation
                 model.pushStackElement();
-                reference.dispose();
-            }));
+                editor.focus();
```

Comment:
I am not sure what will happen if an editor is not visible. You can also pass `revealIfVisible` to `open`.

Reply:
I will remove the `focus` completely. I don't think applying an edit should do anything the focus. E.g. you do a complex refactoring, you want to stay where you are.